### PR TITLE
Eagerly dereference resources

### PR DESCRIPTION
**Pull request by bhhari**
```
== NO RELEASE NOTE ==
```
These changes will speed up the GC process for these variables.

### CODE REVIEW

File: presto-orc/src/main/java/com/facebook/presto/orc/OrcSelectiveRecordReader.java [hunk 1]
```diff
@@ -849,14 +849,6 @@ private void initializeOutputPositions(int positionCount)
     public void close()
             throws IOException
     {
-        try (Closer closer = Closer.create()) {
```
Comment (bhhari) [inline presto-orc/src/main/java/com/facebook/presto/orc/OrcSelectiveRecordReader.java:852]: this is handled in super.close()

File: presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionarySelectiveStreamReader.java [hunk 2]
```diff
@@ -278,13 +278,14 @@ private void openRowGroup()
         if (!dictionaryOpen && dictionarySize > 0) {
             if (dictionary.length < dictionarySize) {
                 dictionary = new long[dictionarySize];
+                systemMemoryContext.setBytes(sizeOf(dictionary));
```
Comment (mbasmanova) [inline presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionarySelectiveStreamReader.java:281]: This change doesn't seem right and is not part of nulling member variables in close.

File: presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionarySelectiveStreamReader.java [hunk 3]
```diff
@@ -278,13 +278,14 @@ private void openRowGroup()
...
             }
 
             LongInputStream dictionaryStream = dictionaryDataStreamSource.openStream();
             if (dictionaryStream == null) {
                 throw new OrcCorruptionException(streamDescriptor.getOrcDataSourceId(), "Dictionary is not empty but data stream is not present");
             }
-            dictionaryStream.nextLongVector(dictionarySize, dictionary);
+            dictionaryStream.next(dictionary, dictionarySize);
```
Comment (mbasmanova) [inline presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionarySelectiveStreamReader.java:288]: What's the reason for this change? Perhaps, move it into a separate commit and explain it in commit messase.

File: presto-spi/src/main/java/com/facebook/presto/spi/block/ClosingBlockLease.java [hunk 4]
```diff
@@ -65,6 +65,7 @@ public void close()
             return;
         }
         closed = true;
+        block = null;
```
Comment (mbasmanova) [inline presto-spi/src/main/java/com/facebook/presto/spi/block/ClosingBlockLease.java:68]: This change is problematic. The lease doesn't own the block and the same block can be used to create multiple leases.

Comment (bhhari) [reply presto-spi/src/main/java/com/facebook/presto/spi/block/ClosingBlockLease.java:68]: All the usage seems to create a new block everytime before getting the lease

Comment (mbasmanova) [reply presto-spi/src/main/java/com/facebook/presto/spi/block/ClosingBlockLease.java:68]: @bhhari Actually, I got confused. This is fine as lease can not be used after "close()".

File: presto-orc/src/main/java/com/facebook/presto/orc/reader/MapDirectSelectiveStreamReader.java [hunk 5]
```diff
@@ -658,6 +658,26 @@ public String toString()
     @Override
     public void close()
     {
+        if (keyReader != null) {
+            keyReader.close();
```
Comment (mbasmanova) [inline presto-orc/src/main/java/com/facebook/presto/orc/reader/MapDirectSelectiveStreamReader.java:662]: @bhhari do you need to add `keyReader = null;`? same for valueReader
